### ISSUE TITLE
Handle restarts in fixturenet-eth stack

### ISSUE THREAD

**Issue by prathamesh0** [cerc-io/stack-orchestrator#303]
Currently, the containers in `fixturenet-eth` [stack](https://github.com/cerc-io/stack-orchestrator/tree/main/app/data/stacks/fixturenet-eth) start afresh on a restart (`deploy down`, then `deploy up`), reinitializing the nodes and mining blocks from 0.

For it's usage in other stacks (eg. [fixturenet-optimism](https://github.com/cerc-io/stack-orchestrator/tree/main/app/data/stacks/fixturenet-optimism)), where we may have to stop the stack for some reason, we need it to save the state and pickup from where it left off on doing a restart.

**Comment by prathamesh0** [comment]
Trying to handle restarts by using mounted volumes for the data directories of geth and lighthouse nodes in #324
Although not facing an issue with geth nodes, the lighthouse nodes aren't able to restart properly.

Relevant logs from the concerned processes:
geth node:
```bash
DEBUG[04-17|06:36:31.409] Served eth_blockNumber                   conn=192.168.16.7:52554 reqid=1 duration="100.66µs"
DEBUG[04-17|06:36:36.945] Served eth_chainId                       conn=192.168.16.7:46070 reqid=1 duration="102.309µs"
DEBUG[04-17|06:36:36.946] Served eth_getBlockByNumber              conn=192.168.16.7:46070 reqid=1 duration="308.362µs"
DEBUG[04-17|06:36:37.136] Served eth_syncing                       conn=192.168.16.7:46084 reqid=1 duration="28.795µs"
INFO [04-17|06:36:37.136] Ignoring beacon update to old head       number=20 hash=8fec43..e37eb7 age=1m45s   have=31
DEBUG[04-17|06:36:37.136] Served engine_forkchoiceUpdatedV1        conn=192.168.16.7:46088 reqid=1 duration="168.303µs"
DEBUG[04-17|06:36:37.136] Served engine_exchangeTransitionConfigurationV1 conn=192.168.16.7:46092 reqid=1 duration="123.211µs"
INFO [04-17|06:36:37.136] Ignoring beacon update to old head       number=20 hash=8fec43..e37eb7 age=1m45s   have=31
DEBUG[04-17|06:36:37.136] Served engine_forkchoiceUpdatedV1        conn=192.168.16.7:46084 reqid=1 duration="133.062µs"
DEBUG[04-17|06:36:37.136] Served eth_syncing                       conn=192.168.16.7:46084 reqid=1 duration="20.889µs"
DEBUG[04-17|06:36:40.138] Served eth_syncing                       conn=192.168.16.7:46084 reqid=1 duration="89.404µs"
INFO [04-17|06:36:40.327] Ignoring beacon update to old head       number=20 hash=8fec43..e37eb7 age=1m48s   have=31
DEBUG[04-17|06:36:40.327] Served engine_forkchoiceUpdatedV1        conn=192.168.16.7:46084 reqid=1 duration="310.654µs"
```

beacon node:
```bash
Apr 17 06:36:55.144 DEBG Execution engine online                 service: exec
Apr 17 06:36:55.144 DEBG Execution engine upcheck complete       state: Synced, service: exec
Apr 17 06:36:55.199 ERRO No valid eth1_data votes, `votes_to_consider` empty, outcome: casting `state.eth1_data` as eth1 vote, genesis_time: 1681713001, earliest_block_timestamp: None, lowest_block_number: None, service: deposit_contract_rpc
Apr 17 06:36:55.199 DEBG Produced vote for eth1 chain            block_hash: 0x4242424242424242424242424242424242424242424242424242424242424242, deposit_count: 80, deposit_root: 0x41da05b85b39025eff3a7e6ef714e73240b39ca1f4b2b716701bcc930117c878, service: deposit_contract_rpc
Apr 17 06:36:55.199 DEBG Issuing engine_getPayload               parent_hash: 0x8fec4395d511e6fe3711dc1df762846e5855e00fc2f2baf2ec88f60679e37eb7, timestamp: 1681713415, prev_randao: 0x7cfbf7bf925b05751d41cecbd453b51ceabf0063083d73a5974b6e957d57f90b, suggested_fee_recipient: 0xe6ce22afe802caf5ff7d3845cec8c736ecc8d61f, service: exec
Apr 17 06:36:55.201 ERRO Exec engine unable to produce payload   status: PayloadStatusV1 { status: Valid, latest_valid_hash: Some(0x8fec4395d511e6fe3711dc1df762846e5855e00fc2f2baf2ec88f60679e37eb7), validation_error: None }, msg: No payload ID, the engine is likely syncing. This has the potential to cause a missed block proposal., service: exec
Apr 17 06:36:55.201 ERRO Execution engine call failed            error: PayloadIdUnavailable, service: exec
```

validator client:
```bash
Apr 17 06:36:37.703 INFO Enabled validator                       voting_pubkey: 0x872c61b4a7f8510ec809e5b023f5fdda2105d024c470ddbbeca4bc74e8280af0d178d749853e8f6a841083ac1b4db98f, sig
ning_method: local_keystore
Apr 17 06:36:37.703 WARN Ignored stale lockfile                  cause: Ungraceful shutdown (harmless) OR non-Lighthouse client using this keystore (risky), path: ../build/cl/node_1/v
alidators/0x872c61b4a7f8510ec809e5b023f5fdda2105d024c470ddbbeca4bc74e8280af0d178d749853e8f6a841083ac1b4db98f/voting-keystore.json.lock
Apr 17 06:36:37.703 INFO Enabled validator                       voting_pubkey: 0x9893413c00283a3f9ed9fd9845dda1cea38228d22567f9541dccc357e54a2d6a6e204103c92564cbc05f4905ac7c493a, sig
ning_method: local_keystore
Apr 17 06:36:37.703 WARN Ignored stale lockfile                  cause: Ungraceful shutdown (harmless) OR non-Lighthouse client using this keystore (risky), path: ../build/cl/node_1/v
alidators/0x9893413c00283a3f9ed9fd9845dda1cea38228d22567f9541dccc357e54a2d6a6e204103c92564cbc05f4905ac7c493a/voting-keystore.json.lock
.
.
.
Apr 17 06:36:40.335 CRIT Error whilst producing block            message: Recoverable("Some endpoints failed, num_failed: 1 http://localhost:8001/ => RequestFailed(Recoverable(\"Error from beacon node when producing block: ServerMessage(ErrorMessage { code: 500, message: \\\"UNHANDLED_ERROR: GetPayloadFailed(EngineError(Api { error: PayloadIdUnavailable }))\\\", stacktraces: [] })\"))"), service: block
```

However, when I tried running only geth in docker and lighthouse nodes locally, restarting nodes worked as expected (tried restarting just the lighthouse nodes as well as both geth (uses volumes for datadirs) and lighthouse nodes).
The errors/warnings from the logs above didn't appear in this configuration.

**Comment by prathamesh0** [comment]
Since the beacon node and validator client are different processes (running in the same container in background) with their data in different directories, tried setups where only one of them is running locally to further narrow down the issue.
Restarts worked when the beacon node was running locally with everything else (including the validator client) run using docker. So it appears that the problem is with the beacon node / it's configuration.

Also, getting the same behaviour when using the latest lighthouse image (`sigp/lighthouse:v4.0.1-modern` instead of `sigp/lighthouse:v3.2.1-modern`) (there was a hardfork recently and our geth has been upgraded for that)

**Comment by prathamesh0** [comment]
The issues mentioned above were a result of lighthouse processes (beacon node and validator) being abruptly terminated on a `deploy down`; have implemented the graceful shutdowns in #324

**Comment by prathamesh0** [comment]
Currently, on a restart, the peers in two lighthouse nodes don't connect to each other.

Logs from the second lighthouse node:
```bash
Apr 18 07:22:30.865 DEBG Discovery query yielded no results.     service: libp2p
```

Expected:
```bash
Apr 12 06:05:45.353 DEBG Discovery query completed               peers_found: 1, service: libp2p
Apr 12 06:05:45.353 DEBG Starting a new peer discovery query     wanted: 1, outbound: 0, target: 1, connected: 0, service: libp2p
Apr 12 06:05:45.353 DEBG Dialing discovered peer                 peer_id: 16Uiu2HAmSdLEQcqRy4gP148Y6NATotujXzaLfJ8PNjZ8KjEjpEtH, service: libp2p
```
However, this doesn't hinder the block production on a restart.
To be investigated.
